server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -206,6 +206,12 @@ public void allocate(RoutingAllocation allocation, ActionListener<Void> listener
         assert MasterService.assertMasterUpdateOrTestThread() : Thread.currentThread().getName();
         assert allocation.ignoreDisable() == false;
 
+        if (allocation.routingTable().indicesRouting().isEmpty()) {
+            logger.debug("No allocation needed for empty routing table");
+            listener.onResponse(null);
+            return;
```

Comment:
I am wondering about a case when we create and then delete index.
Would it (1) remove it from the desired balance and (2) reconcile the cluster cluster state to unallocate it?

Reply:
Sorry for my ignorance in this area. Do we need (2)? Deleting an index [removes](https://github.com/elastic/elasticsearch/blob/c1c59eb41d6864c3c4d95b01629ad6b72525e9a4/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java#L192) it from the routingTable. So we would have computed an empty desiredBalance which is then [short-circuited](https://github.com/elastic/elasticsearch/blob/c1c59eb41d6864c3c4d95b01629ad6b72525e9a4/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java#L135-L138) in Reconciler. It does _not_ seem to do anything actually useful. If we want the empty desiredBalance, we can also directly set it here? 

The need for fixing the test failure is to handle state-not-recovered and empty routingTable is a proxy to it since we don't have access to cluster state at this point. Therefore I chose to check for empty routingTable as it seems safe to me. But please let me know if there are consequences that I am not aware of.

Reply:
Yeah, you are correct about removing an index (it will also be picked by https://github.com/elastic/elasticsearch/blob/a59c182f9f7e9d1bf3d6eecbc0e44f24ff91d053/server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java#L300).

I think it is worth setting empty desired balance here for completeness (for example to make sure we do not report deleted indices in desired balance related apis).

Reply:
On a second thought, I changed it to skip only the [eager reconcile](https://github.com/elastic/elasticsearch/blob/4e17c61d39c64f2522d8ed6245c9c43d41a0dad0/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L219) call which is the cause for the CI failure. This change brought the impact down to minimal since it relies on the existing code flow to update DesiredBalance and the subsequent reconcile (should be short-circuited as we discussed). This works better because we avoid mutating `DesiredBalance` in places other than the single-threaded computation and during master failover, i.e. less racy.

Please see 44d22ddb0892f4fbaadbc9056fc1c8f8a06f2ec0

Reply:
@idegtiarenko I am going to merge this PR since I'd like to close the CI failure issue before the iteration closes. Happy to take any post-merge comment. Thanks!

Reply:
Sorry, forgot to mention that I asked @nicktindall to review the latest change and he has approved it.